### ISSUE TITLE
[Feature Request] Comments

### ISSUE THREAD

**Issue by AlexanderNenninger** [dmilford/rust-3d-demo#2]
Hi Doug,

I followed Your YT tutorial on this project and really liked it. But looking back at the code, in certain places, I have no clue, what each piece is exactly doing. Especially in the `programs::graph_3d` module, since some buffers are initialized in the `render` Function, some in the constructor.

Would it be possible to add a few comments here or there?

**Comment by dmilford** [comment]
Hi Alexander!

Yeah, the comments are pretty skimpy.  My bad :(  Maybe let me know what sections you're interested in and I might pop some in there.

Regarding the buffers in programs::graph_3d... all buffers are "created" within the constructor.  I "fill" some of those buffers like the "XZ" plane (aka position_buffer) within the constructor because those values never, ever change.  For the "Y" values, I buffer those every render because they WILL change continuously on each render.  That's why the position_buffer is a "STATIC_DRAW" and the y_buffer is "DYNAMIC_DRAW".  Things that need to be buffered once and only once, I fill in the constructor.  Things that change over time, I fill in the render.

If I had designed it such that the size of the N x N graph could change dimensions over time (aka change to M x M), then everything would have to go down in the render.  I'd still "create" the buffers in the constructor like I do now, but would fill them in the render.

It all comes down to efficiency.  If something isn't going to change, why re-buffer it every render?

Hope that makes sense.  Let me know if you have any other questions and I'll try to help you out.

Doug

**Comment by AlexanderNenninger** [comment]
Hi Doug,   
thanks for the speedy reply! After looking over Your code, I understand a little better now what's happening in general, although the devil is in the detail:

```
// Make a grid, no magic here
let positions_and_indices = cf::get_position_grid_n_by_n(super::super::constants::GRID_SIZE);
 // Request memory, that's available to JS?
let memory_buffer = wasm_bindgen::memory()
    .dyn_into::<WebAssembly::Memory>()
    .unwrap()
    .buffer();
// Get a pointer to the vertices, but why can we divide by 4? Would't it point to a completely different memory location?
let vertices_location = positions_and_indices.0.as_ptr() as u32 / 4;
// Is vert_array just an view of positions_and_indices.0?
let vert_array = js_sys::Float32Array::new(&memory_buffer).subarray(
    vertices_location,
    vertices_location + positions_and_indices.0.len() as u32,
);
// Try requesting memory on GPU, on failure panic? 
let buffer_position = gl.create_buffer().ok_or("failed to create buffer").unwrap();
// What does this do?
gl.bind_buffer(GL::ARRAY_BUFFER, Some(&buffer_position));
gl.buffer_data_with_array_buffer_view(GL::ARRAY_BUFFER, &vert_array, GL::STATIC_DRAW);
```

Mozilla's documentation on those last two lines is very thin...

-Alex

**Comment by dmilford** [comment]
**Request memory, that's available to JS?**
    Yes, request a memory buffer.

**Is vert_array just an view of positions_and_indices.0?**
    Yes

**Try requesting memory on GPU, on failure panic?** 
    Hence the name, yes, if it can get memory, then great.  Otherwise, panic.

**What does gl.bind_buffer do?**
    Tells WebGL you want to "point" to a buffer as you're preparing to shove data into it
    GL::ARRAY_BUFFER means your data is vertex information, not index information.

**What does gl.buffer_data_with_array_buffer do?**
    Actually shove the data into the buffer that you're currently pointing to.
    GL::STATIC_DRAW means the data is not expected to change even after many renders.
    
Regarding why we divide by 4, I myself was a little puzzled when I first saw code like that... and honestly still am a bit.  I found the following answer from a comment posted by TheRealZerve on the second video.  It has a link to the following github issue.

https://github.com/rustwasm/wasm-bindgen/issues/1438

FROM GITHUB ISSUE
_"If I understand this correctly, this code creates a Float32Array from all of the memory allocated to wasm, and subarray's arguments are indices into the array. Since floats are 4 bytes, it divides by 4 to compute the index into the array from the memory address.

The problem seems to be in the line let arr_location: u32 = indices.as_ptr() as u32 / 4;. Since WebGL indices are 2 bytes, this should divide by 2 instead of 4."_

If you'll notice, I divide by 4 for the positions data and divide by 2 on the indices.

Hope that helps.

Doug
